src/inmanta/references.py

```diff
@@ -0,0 +1,450 @@
+        return resource
+
+
+ArgumentTypes = typing.Annotated[
+    LiteralArgument | ReferenceArgument | GetArgument | PythonTypeArgument | ResourceArgument,
+    pydantic.Field(discriminator="type"),
+]
+""" A list of all specific types of arguments. Pydantic uses this to instantiate the correct argument class
+"""
+
+
+class BaseModel(pydantic.BaseModel):
+    """A base model class for references and mutators"""
+
+    type: ReferenceType
+    args: list[ArgumentTypes]
+
+
+class ReferenceModel(BaseModel):
+    """A reference"""
+
+    id: uuid.UUID
+
+
+class MutatorModel(BaseModel):
+    """A mutator"""
+
+
+C = typing.TypeVar("C", bound="Base")
+
+
+class Base:
+    """A base class for references and mutators"""
+
+    type: typing.ClassVar[ReferenceType]
+
+    def __init__(self, **kwargs: object) -> None:
+        self._arguments: typing.Mapping[str, object] = kwargs
+        self._model: typing.Optional[BaseModel] = None
+        # TODO: do we want to enforce type correctness when creating a reference? This also has impact on how arguments are
+        #       are serialized: based on instance types or on static types
+
+    @classmethod
+    def deserialize(
+        cls: typing.Type[C],
+        ref: BaseModel,
+        resource: "inmanta.resources.Resource",
+    ) -> C:
+        """Deserialize the reference or mutator.
+
+        :param ref: The model of the reference to deserialize
+        :param resource: The resource to use as a read-only reference
+        """
+        return cls(**{arg.name: arg.get_arg_value(resource) for arg in ref.args})
+
+    @abc.abstractmethod
+    def serialize(self) -> BaseModel:
+        """Serialize to be able to add them to a resource"""
+
+    def serialize_arguments(self) -> Tuple[uuid.UUID, list[ArgumentTypes]]:
+        """Serialize the arguments to this class"""
+        parameters = inspect.get_annotations(type(self).__init__, eval_str=True)
+        arguments: list[ArgumentTypes] = []
+        for name, value in self._arguments.items():
+            param_type = parameters.get(name)
+            match name, value, param_type:
+                case _, str() | int() | float() | bool(), _:
+                    arguments.append(LiteralArgument(name=name, value=value))
+
+                case _, Reference(), _:
+                    model = value.serialize()
+                    arguments.append(ReferenceArgument(name=name, id=model.id))
+
+                case "resource", _, inmanta.resources.Resource:
+                    arguments.append(ResourceArgument(name=name))
+
+                case _, type(), _ if value in [str, float, int, bool]:
+                    arguments.append(PythonTypeArgument(name=name, value=value.__name__))
+
+                case _:
+                    raise TypeError(f"Unable to serialize argument `{name}` of `{self}`")
+
+        data = json.dumps({"type": self.type, "args": arguments}, default=util.api_boundary_json_encoder, sort_keys=True)
```

Comment:
may not be entirely stable, because the argument could be re-ordered when manipulating` _arguments`